### PR TITLE
Support generators and unrolling in while

### PR DESCRIPTION
**Pull request by josevalim**

### CODE REVIEW

File: nx/lib/nx/defn/kernel.ex [hunk 1]
```diff
@@ -1046,42 +1046,79 @@ defmodule Nx.Defn.Kernel do
...
+
+      while acc = 0, i <- tensor, unroll: true do
+        acc + i
+      end
+
+  Or unroll it in batches:
+
+      while acc = 0, i <- tensor, unroll: 4 do
+        acc + i
+      end
```
Comment (jonatanklosko) [inline nx/lib/nx/defn/kernel.ex:1080]:
```suggestion
      end
  
  Unrolling means that the the `while` body is automatically duplicated
  a certain amount of times, as if you wrote all iterations by hand. This
  makes the final expression larger, which causes a longer compilation
  time, however it enables additional compile-time optimizations (such as
  fusion), improving the runtime efficiency.
```

File: nx/lib/nx/defn/kernel.ex [hunk 2]
```diff
@@ -1046,42 +1046,79 @@ defmodule Nx.Defn.Kernel do
...
```
Comment (jonatanklosko) [inline nx/lib/nx/defn/kernel.ex:1081]: I think we should also add a warning note that `while` should generally be avoided unless there is no way to express the algorithm with regular tensor operations.

File: nx/lib/nx/defn/expr.ex [hunk 3]
```diff
@@ -305,18 +305,99 @@ defmodule Nx.Defn.Expr do
...
-        {parameter(expr, context, counter), {counter + 1, merge_context!(expr, acc)}}
-      end)
+    case generator do
+      :none ->
+        if opts != [] do
+          raise CompileError,
+            line: line,
+            file: file,
+            description:
+              "options are only supported on while with conditions, only with generators, got: #{inspect(opts)}"
```
Comment (jonatanklosko) [inline nx/lib/nx/defn/expr.ex:318]: ?

```suggestion
              "options are only supported on while with generators, got: #{inspect(opts)}"
```

Comment (josevalim) [reply nx/lib/nx/defn/expr.ex:318]:
```suggestion
              "options are not supported on while with conditions, only with generators, got: #{inspect(opts)}"
```

File: nx/lib/nx/defn/kernel.ex (hunk 2 above)
Comment (josevalim) [reply nx/lib/nx/defn/kernel.ex:1085]: Done!

**Pull request merged**